### ISSUE TITLE
Add image as data url?

### ISSUE THREAD

**Issue by danielbsig** [fraywing/textAngular#1177]
Just a question: I know that other wysiwyg editors have the option to allow images to be added as data urls, thereby removing the need to upload them separately. Has this option been considered in this library? Or has it perhaps been implemented, but I simply can't find it?

**Comment by pbassut** [comment]
You mean input an image as base64 or have textangular download the image and add it as base64?
Neither of them is implemented. 
It's perfectly possible, though.

**Comment by danielbsig** [comment]
Input an images as base64. I suppose it will be relatively easy to add this as an extenstion to the library, but having it in the library itself would of course be more convenient. Just a simple matter of figuring out the ui, using an <input type="file"> to select the image would be a challenge to make it look in line with the rest of the toolbar buttons.

**Comment by pbassut** [comment]
Well, I don't see how messing with base64 would be useful from the user-side. Nobody, in a sane mind, would pass around base64 strings to represent images. Either they would hand a url or the file itself.
So, adding that as a tool for your particular situation would be more convenient, I'd say.

**Comment by danielbsig** [comment]
I would say that there are very good points for working with images as base64 encoded text. That way, an image added to a text block can be stored with the text block, and not stored externally. Links tend to break, for instance, what happens if the images are moved, or even just if the urls are changed slightly? Yes I'm aware of redirects, but that is not always feasible or even something people remember to do. Also, if the text block is copied, and one of them is later edited, can the user delete the original image from the hosted location? Users will either delete them when they shouldn't, thereby breaking the other text block, or not delete them when they should, resulting in build up of stale images, taking up space there.

In certain cases, it would of course make more sense to link to images. But having the option to choose is definitely better than having either method selected for you by someone else.

These are my 2 cents in any case. I can hear that you are not overly fond of the idea, so I'm not expecting you to implement this. I'm just saying having an open mind to the idea would not be a bad idea. I'll see if I can submit a PR for this, although I won't promise anything.

**Comment by pbassut** [comment]
You made it look like a user would actually input base64 data. Which I think is not what you intent to do.
Instead you want textAngular to fetch the image from the url and use it's base64 version instead.

As you said, if you use external resources very much(like when writing an articles for a blog or something) things will likely break at some point because, again as you pointed out, the url could return a 404 or the image changed completely. 

About the idea of fetching the image and using it's base64 version I can say I'm willing to accept this. Could be a flag like 'useImageBase64'.

Still, it's not a feature that 90% of the users would use. Just my personal opinion.

**Comment by danielbsig** [comment]
Yes, sorry for the confusion. The users would always just be presented with an UI to select a file (FileOpen dialog, dragdrop, whatever...) and the library would then use the FileReader API to access the base64 representation of the image and add it to the text field.

And yes, I believe that this should be a separate button, completely independent from the one already present.

Obviously I have no idea how many users use this. I'm fairly sure that I would use it, and we would just have to see what other users/developers would do.

**Comment by pbassut** [comment]
Well, if it doesn't conflict with the normal use flow of the library, I'm sure it's a PR I would accept. Because then, when somebody needs they would just enable it, right.

**Comment by outreal** [comment]
Hi,
Please let me join your discussion.I would be very happy to be able to load images and have them converted to base64, in a <img src="data:image/png;base64,[data]>. That's what summernote does, and it's really convenient, specially for emails.

So with @danielbsig , that makes two of us :-)

And let me thank you for textAngular, which is a great piece of software!
